docs/source/SETTINGS.rst
========
Settings
========

Exporting / Importing
=====================

The settings make use of the pydantic-settings_ library, following examples show how it can be used to export and import the settings to a JSON files.

Writing settings to a JSON file:

.. code-block:: python

    import json
    from aioslsk.settings import Settings, CredentialsSettings

    settings = Settings(
        credentials=CredentialsSettings(username='user', password='testpass')
    )
    settings_filename = 'my_settings.json'

    with open(settings_filename, 'w') as fh:
        fh.write(settings.dump_model(mode='json'))


Reading settings from a JSON file:

.. code-block:: python

    import json
    from aioslsk.settings import Settings, CredentialsSettings

    settings_filename = 'my_settings.json'

    with open(settings_filename, 'r') as fh:
        py_settings = json.load(fh)

    loaded_settings = Settings(**py_settings)


Settings
========

This is a full list of settings used by the library

Credentials
-----------

+------------------------------+--------+------------------------------------------------------------------------+-----------+
|          Parameter           |  Type  |                              Description                               |  Default  |
+==============================+========+========================================================================+===========+
| credentials.username         | string | Username used to login                                                 | <not set> |
+------------------------------+--------+------------------------------------------------------------------------+-----------+
| credentials.password         | string | Password used to login                                                 | <not set> |
+------------------------------+--------+------------------------------------------------------------------------+-----------+
| credentials.info.description | string | Personal description, will be returned when a peer request info on you | <not set> |
+------------------------------+--------+------------------------------------------------------------------------+-----------+
| credentials.info.picture     | string | Picture, will be returned when a peer request info on you              | <not set> |
+------------------------------+--------+------------------------------------------------------------------------+-----------+


Network
-------

+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
|             Parameter              |  Type   |                                            Description                                            |      Default       |
+====================================+=========+===================================================================================================+====================+
| network.server.hostname            | string  |                                                                                                   | server.slsknet.org |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.server.port                | integer |                                                                                                   | 2416               |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.server.reconnect.auto      | boolean | Automatically try to reconnect to the server when the server disconnects. No attempt to reconnect | false              |
|                                    |         | will be made when: disconnecting, server closed connection, credentials are not set               |                    |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.server.reconnect.timeout   | integer | Timeout after which we should try reconnecting to the server                                      | 10                 |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.listening.port             | integer | Port to listen for other peers (clear)                                                            | 61000              |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.listening.obfuscated_port  | integer | Port to listen for other peers (obfuscated)                                                       | 61001              |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.listening.error_mode       | string  | Defines when an error should be raised when connecting the listening connection (all, any, clear) | clear              |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.peer.obfuscate             | boolean | When true, prefer obfuscated connections as much as possible                                      | false              |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.peer.connect_mode          | string  | Specifies how to make peer connections (race, fallback)                                           | race               |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.upnp.enabled               | boolean | Automatically configure UPnP for the listening ports                                              | false              |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.upnp.lease_duration        | integer | Length of the UPnP port mapping lease duration (in seconds)                                       | 0                  |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.upnp.check_interval        | integer | Time between checking UPnP mappings (in seconds)                                                  | 600                |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.upnp.search_timeout        | integer | Maximum time to search for UPnP devices (in seconds)                                              | 10                 |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.limits.upload_speed_kbps   | integer | Upload speed limit in kbps (0 = no limit)                                                         | 0                  |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+
| network.limits.download_speed_kbps | integer | Download speed limit in kbps (0 = no limit)                                                       | 0                  |
+------------------------------------+---------+---------------------------------------------------------------------------------------------------+--------------------+


Sharing
-------

+----------------------+----------------------------------------------+---------------------------------------------------------------------------+-----------------------------+
|      Parameter       |                     Type                     |                                Description                                |           Default           |
+======================+==============================================+===========================================================================+=============================+
| shares.download      | string                                       | Directory to which files will be downloaded to                            | <current working directory> |
+----------------------+----------------------------------------------+---------------------------------------------------------------------------+-----------------------------+
| shares.directories   | array[:class:`.SharedDirectorySettingEntry`] | List of shared directories (see structure for each entry below)           | <empty>                     |
+----------------------+----------------------------------------------+---------------------------------------------------------------------------+-----------------------------+
| shares.scan_on_start | boolean                                      | Schedule a scan as soon as the client starts up (after reading the cache) | true                        |
+----------------------+----------------------------------------------+---------------------------------------------------------------------------+-----------------------------+


The ``shares.directories`` list contains objects which have the following parameters (object of :class:`.SharedDirectorySettingEntry`):

+------------+---------------+-----------------------------------------------------+-----------+
| Parameter  |     Type      |                     Description                     | Mandatory |
+============+===============+=====================================================+===========+
| path       | string        | Maximum amount of simultaneously uploads            | yes       |
+------------+---------------+-----------------------------------------------------+-----------+
| share_mode | string        | Possible values: `everyone`, `friends`, `users`     | yes       |
+------------+---------------+-----------------------------------------------------+-----------+
| users      | array[string] | List of specific users to share this directory with | no        |
+------------+---------------+-----------------------------------------------------+-----------+


Users / Rooms
-------------

+----------------------------+-------------------------------------+----------------------------------------------------------------------------------------+---------+
|         Parameter          |                Type                 |                                      Description                                       | Default |
+============================+=====================================+========================================================================================+=========+
| rooms.auto_join            | boolean                             | Automatically rejoin rooms when logon is successful                                    | true    |
+----------------------------+-------------------------------------+----------------------------------------------------------------------------------------+---------+
| rooms.private_room_invites | boolean                             | Enable or disable private rooms invitations                                            | true    |
+----------------------------+-------------------------------------+----------------------------------------------------------------------------------------+---------+
| rooms.favorites            | array[string]                       | List of rooms that will automatically be joined                                        | <empty> |
+----------------------------+-------------------------------------+----------------------------------------------------------------------------------------+---------+
| users.friends              | array[string]                       | List users considered friends                                                          | <empty> |
+----------------------------+-------------------------------------+----------------------------------------------------------------------------------------+---------+
| users.blocked              | map[string, :class:`.BlockingFlag`] | List of blocked users. Key indicates the username, value is one or more blocking flags | <empty> |
+----------------------------+-------------------------------------+----------------------------------------------------------------------------------------+---------+


Interests
---------

+----------------------------+---------------+-----------------------------------------------------+---------+
|         Parameter          |     Type      |                     Description                     | Default |
+============================+===============+=====================================================+=========+
| interests.liked            | array[string] | List of liked interests                             | <empty> |
+----------------------------+---------------+-----------------------------------------------------+---------+
| interests.hated            | array[string] | List of hated interests                             | <empty> |
+----------------------------+---------------+-----------------------------------------------------+---------+


Search
------

+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
|               Parameter                |     Type      |                                                                           Description                                                                           | Default |
+========================================+===============+=================================================================================================================================================================+=========+
| searches.receive.max_results           | integer       | Maximum amount of search results returned when replying to search requests from other peers                                                                     | 100     |
+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| searches.receive.store_amount          | integer       | Amount of received searches to store in the client                                                                                                              | 500     |
+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| searches.send.store_results            | boolean       | Whether to store search results internally                                                                                                                      | true    |
+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| searches.send.request_timeout          | integer       | Timeout for sent search requests, when the timeout is reached the request will be removed and search results will no longer be accepted (0 = keep indefinitely) | 0       |
+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| searches.send.wishlist_request_timeout | integer       | Timeout for sent wishlist requests (0 = keep indefinitely, -1 = use the interval advertised by the server)                                                      | -1      |
+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+
| searches.wishlist                      | array[object] | List of wishlist items. Object definition is defined below                                                                                                      | <empty> |
+----------------------------------------+---------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------+---------+


Following object should be used for ``searches.wishlist`` object:

+-----------+---------+-------------------------------------------+---------+
| Parameter |  Type   |                Description                | Default |
+===========+=========+===========================================+=========+
| query     | string  | Wishlist item search query                | <empty> |
+-----------+---------+-------------------------------------------+---------+
| enabled   | boolean | Whether this query is enabled or disabled | true    |
+-----------+---------+-------------------------------------------+---------+


Transfers
---------

+-------------------------------+---------+-------------------------------------------------+---------+
|           Parameter           |  Type   |                   Description                   | Default |
+===============================+=========+=================================================+=========+
| transfers.limits.upload_slots | integer | Maximum amount of simultaneously uploads        | 2       |
+-------------------------------+---------+-------------------------------------------------+---------+
| transfers.report_interval     | float   | Transfer progress reporting interval in seconds | 0.250   |
+-------------------------------+---------+-------------------------------------------------+---------+


Debug
-----

+----------------------------+---------------------+----------------------------------------------------+---------+
|         Parameter          |        Type         |                    Description                     | Default |
+============================+=====================+====================================================+=========+
| debug.search_for_parent    | boolean             | Toggle searching for a distributed parent          | true    |
+----------------------------+---------------------+----------------------------------------------------+---------+
| debug.user_ip_overrides    | map[string, string] | Mapping of username and IP addresses, overrides    | <empty> |
+----------------------------+---------------------+----------------------------------------------------+---------+
| debug.log_connection_count | boolean             | Periodically log the amount of current connections | false   |
+----------------------------+---------------------+----------------------------------------------------+---------+


.. _pydantic-settings: https://docs.pydantic.dev/latest/concepts/pydantic_settings/


src/aioslsk/settings.py
from collections.abc import Set
import os
from typing import Annotated, Any, Optional, Sequence

from pydantic import BaseModel, BeforeValidator, Field, model_validator
from pydantic_settings import BaseSettings

from .constants import (
    UPNP_DEFAULT_CHECK_INTERVAL,
    UPNP_DEFAULT_LEASE_DURATION,
    UPNP_DEFAULT_SEARCH_TIMEOUT,
)
from .network.network import ListeningConnectionErrorMode, PeerConnectMode
from .shares.model import DirectoryShareMode
from .user.model import BlockingFlag


def translate_blocked_users(value: Any) -> dict[str, BlockingFlag]:
    """Converts blocked user set to a dictionary object"""
    if isinstance(value, dict):
        return value

    if isinstance(value, (Sequence, Set)):
        blocked_dict = {}
        for username in value:
            blocked_dict[username] = BlockingFlag.ALL

        return blocked_dict

    raise ValueError(f"unable to translate type {type(value)}")


class SharedDirectorySettingEntry(BaseModel, validate_assignment=True):
    path: str
    share_mode: DirectoryShareMode = DirectoryShareMode.EVERYONE
    users: list[str] = Field(default_factory=list)


class WishlistSettingEntry(BaseModel, validate_assignment=True):
    query: str
    enabled: bool = True


class UpnpSettings(BaseModel, validate_assignment=True):
    enabled: bool = True
    lease_duration: int = UPNP_DEFAULT_LEASE_DURATION
    check_interval: int = UPNP_DEFAULT_CHECK_INTERVAL
    search_timeout: int = UPNP_DEFAULT_SEARCH_TIMEOUT


class ReconnectSettings(BaseModel, validate_assignment=True):
    auto: bool = False
    timeout: int = 10


class ServerSettings(BaseModel, validate_assignment=True):
    hostname: str = 'server.slsknet.org'
    port: int = 2416
    reconnect: ReconnectSettings = ReconnectSettings()


class ListeningSettings(BaseModel, validate_assignment=True):
    error_mode: ListeningConnectionErrorMode = ListeningConnectionErrorMode.CLEAR
    port: int = 60000
    obfuscated_port: int = 60001


class PeerSettings(BaseModel, validate_assignment=True):
    obfuscate: bool = False
    connect_mode: PeerConnectMode = PeerConnectMode.RACE


class NetworkLimitSettings(BaseModel, validate_assignment=True):
    upload_speed_kbps: int = 0
    download_speed_kbps: int = 0


class NetworkSettings(BaseModel, validate_assignment=True):
    server: ServerSettings = Field(default_factory=ServerSettings)
    listening: ListeningSettings = Field(default_factory=ListeningSettings)
    peer: PeerSettings = Field(default_factory=PeerSettings)
    upnp: UpnpSettings = Field(default_factory=UpnpSettings)
    limits: NetworkLimitSettings = Field(default_factory=NetworkLimitSettings)


class UserInfoSettings(BaseModel, validate_assignment=True):
    description: Optional[str] = None
    picture: Optional[bytes] = None


class CredentialsSettings(BaseModel, validate_assignment=True):
    username: str
    password: str
    info: UserInfoSettings = Field(default_factory=UserInfoSettings)

    def are_configured(self) -> bool:
        """Returns whether the credentials are correctly configured"""
        return bool(self.username) and self.password is not None


class SearchSendSettings(BaseModel, validate_assignment=True):
    store_results: bool = True
    request_timeout: int = 0
    wishlist_request_timeout: int = -1


class SearchReceiveSettings(BaseModel, validate_assignment=True):
    store_amount: int = 500
    max_results: int = 100


class SearchSettings(BaseModel, validate_assignment=True):
    send: SearchSendSettings = Field(default_factory=SearchSendSettings)
    receive: SearchReceiveSettings = Field(default_factory=SearchReceiveSettings)
    wishlist: list[WishlistSettingEntry] = Field(default_factory=list)

    @model_validator(mode='before')
    @classmethod
    def handle_max_result_alias(cls, values: dict):
        if 'max_results' in values:
            value = values.pop('max_results')
            if 'receive' not in values:
                values['receive'] = {}

            values['receive']['max_results'] = value

        return values

    @property
    def max_results(self) -> int:
        return self.receive.max_results

    @max_results.setter
    def max_results(self, value: int):
        self.receive.max_results = value


class TransferLimitSettings(BaseModel, validate_assignment=True):
    upload_slots: int = 2


class TransfersSettings(BaseModel, validate_assignment=True):
    limits: TransferLimitSettings = Field(default_factory=TransferLimitSettings)
    report_interval: float = 0.250


class SharesSettings(BaseModel, validate_assignment=True):
    scan_on_start: bool = True
    download: str = os.getcwd()
    directories: list[SharedDirectorySettingEntry] = Field(default_factory=list)


class RoomsSettings(BaseModel, validate_assignment=True):
    auto_join: bool = True
    private_room_invites: bool = True
    favorites: set[str] = Field(default_factory=set)


class UsersSettings(BaseModel, validate_assignment=True):
    friends: set[str] = Field(default_factory=set)
    blocked: Annotated[
        dict[str, BlockingFlag],
        BeforeValidator(translate_blocked_users)
    ] = Field(default_factory=dict)

    def is_blocked(self, username: str, flag: BlockingFlag) -> bool:
        return bool(self.blocked.get(username, BlockingFlag.NONE) & flag)


class InterestsSettings(BaseModel, validate_assignment=True):
    liked: set[str] = Field(default_factory=set)
    hated: set[str] = Field(default_factory=set)


class DebugSettings(BaseModel, validate_assignment=True):
    search_for_parent: bool = True
    ip_overrides: dict[str, str] = Field(default_factory=dict)
    log_connection_count: bool = False


class Settings(BaseSettings, validate_assignment=True):
    network: NetworkSettings = Field(default_factory=NetworkSettings)
    credentials: CredentialsSettings = Field(default_factory=CredentialsSettings)  # type: ignore
    searches: SearchSettings = Field(default_factory=SearchSettings)
    shares: SharesSettings = Field(default_factory=SharesSettings)
    users: UsersSettings = Field(default_factory=UsersSettings)
    rooms: RoomsSettings = Field(default_factory=RoomsSettings)
    interests: InterestsSettings = Field(default_factory=InterestsSettings)
    transfers: TransfersSettings = Field(default_factory=TransfersSettings)
    debug: DebugSettings = Field(default_factory=DebugSettings)


def _main(args):
    import json
    import sys

    if args.command == 'generate':
        settings = Settings(
            credentials=CredentialsSettings(
                username=args.username,
                password=args.password
            )
        )

        dump = settings.model_dump(mode='json')
        json.dump(dump, sys.stdout, indent=2)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Setting file tools")
    subparsers = parser.add_subparsers(dest='command')

    gen_parser = subparsers.add_parser('generate')
    gen_parser.add_argument('--username', '-u', default="TestUser")
    gen_parser.add_argument('--password', '-p', default="Pass0")

    args = parser.parse_args()
    _main(args)
